### PR TITLE
Forbid out-of-range `insert` on `Sequence`

### PR DESCRIPTION
**Pull request by niboshi**
Fixes: #6325

### CODE REVIEW

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
@niboshi Could you resolve the conflicts? I will merge this after confirming all tests pass.

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
@niboshi The `insert()` method of the Python's list object can accept an index larger than the length of the list object. Is it OK not to align with the specification of Python's list here?

**Comment by niboshi** [comment]
IMHO Python's specification of `list` was a failure (there may be other opinions though).
I think it's less error-prone to make such case an error.

@beam2d Do you have any opinion?

File: chainer/sequential.py [hunk 1]
```diff
@@ -260,6 +260,12 @@ def extend(self, sequential):
             self.append(layer)
 
     def insert(self, i, layer):
+        n = len(self._layers)
+        if not (-n <= i < (n + 1)):
+            raise IndexError(
+                'Index out of range: {}'.format(i))
```
Comment (mitmul) [inline chainer/sequential.py]: How about additionally suggesting what should be used instead? For example,
```suggestion
                'Index out of range: {}. Please use .append() method instead.'.format(i))
```

Comment (mitmul) [reply chainer/sequential.py]: @niboshi Doesn't it make sense?

Comment (niboshi) [reply chainer/sequential.py]: When user encounteres this error, I think it's unlikely that he's trying to achieve the behavior of `append`.
For example, if a sequence `seq` has 2 elements, if you want to append a new layer, you can write either `seq.append(new_layer)` or `seq.insert(2, new_layer)` and both should work.
The error is caused when trying something like `seq.insert(3, new_layer)`. I think very few people would write this way. It's more likely that the code has flawed index handling.

Comment (mitmul) [reply chainer/sequential.py]: Got it. Thanks for explaining.

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by mitmul** [comment]
Jenkins, test this please

File: tests/chainer_tests/test_sequential.py [hunk 2]
```diff
@@ -599,12 +593,69 @@ def test_flatten(self):
...
+    def test_index_out_of_bounds_empty_positive(self):
+        seq = chainer.Sequential()
+        with pytest.raises(IndexError):
+            seq.insert(1, functions.sin)
+
+    def test_index_out_of_bounds_empty_negative(self):
         seq = chainer.Sequential()
-        x = numpy.ones((2, 3), numpy.float32)
-        with pytest.raises(RuntimeError):
-            seq(x)
```
Comment (hvy) [inline tests/chainer_tests/test_sequential.py]: Seems like this test is irrelevant to this PR. Shouldn't we it?

**Comment by hvy** [comment]
Basically LGTM. I added a comment regarding a deleted test.

**Comment by niboshi** [comment]
@hvy Sorry, I restored the test. PTAL

File: tests/chainer_tests/test_sequential.py [hunk 3]
```diff
@@ -607,4 +601,69 @@ def test_empty_sequential(self):
             seq(x)
 
 
+class SequentialInsert(unittest.TestCase):
```
Comment (hvy) [inline tests/chainer_tests/test_sequential.py]: How about adding tests inserting `Link`s (similar to the bug report https://github.com/chainer/chainer/issues/6325#issue-412548070)? The code paths are different for functions and links.

Comment (hvy) [inline tests/chainer_tests/test_sequential.py]: A nit.
```suggestion
class TestSequentialInsert(unittest.TestCase):
```

Comment (niboshi) [reply tests/chainer_tests/test_sequential.py]: Hmm, code paths are not different after the fix. What do you mean?
The test complexity would be greatly increase which I don't think worthwhile.

Comment (hvy) [reply tests/chainer_tests/test_sequential.py]: The paths are still different if the index is not out of bounds. How about restoring https://github.com/chainer/chainer/pull/6374/files#diff-a3da12ecfbd5149195d80196a0b251c1L424 ?

Comment (toslunar) [reply tests/chainer_tests/test_sequential.py]: `TestSequentialInsert` should test inserting `Link`s, while `Sequential.insert(_, link)` is called elsewhere in the file.  Having both `TestSequential.test_insert` and `TestSequentialInsert` is confusing.

Comment (hvy) [reply tests/chainer_tests/test_sequential.py]: I believe that's a perfectly valid point. My point was that we should still test inserting links, not necessarily with all combinations of invalid indices. We could do that by more or less moving the removed method to this class.

Comment (niboshi) [reply tests/chainer_tests/test_sequential.py]: It would be insufficient to literally restoring the single test.

Comment (niboshi) [reply tests/chainer_tests/test_sequential.py]: Fixed in 5030fb7c304cff0813d391596d1fa0114e436acd.

**Comment by hvy** [comment]
Jenkins, test this please.

**Pull request merged**